presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapNumericKeyBlockWriter.java

```diff
@@ -0,0 +1,247 @@
+package com.facebook.presto.orc.writer;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.ColumnarMap;
+import com.facebook.presto.common.function.SqlFunctionProperties;
+import com.facebook.presto.common.type.MapType;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.orc.ColumnWriterOptions;
+import com.facebook.presto.orc.DwrfDataEncryptor;
+import com.facebook.presto.orc.metadata.ColumnEncoding;
+import com.facebook.presto.orc.metadata.MetadataWriter;
+import com.facebook.presto.orc.metadata.OrcType;
+import com.facebook.presto.orc.metadata.statistics.ColumnStatistics;
+import com.facebook.presto.orc.stream.StreamDataOutput;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+
+import static com.facebook.presto.common.type.TimeZoneKey.UTC_KEY;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Locale.ENGLISH;
+
+public class FlatMapNumericKeyBlockWriter
+        implements FlatMapBlockWriter
+{
+    private final int valueColumnId;
+    private final Type keyType;
+    private final Type valueType;
+    private final OrcType orcKeyType;
+    ColumnWriterOptions columnWriterOptions;
+    private final Optional<DwrfDataEncryptor> dwrfEncryptor;
+    private final MetadataWriter metadataWriter;
+    private final ColumnWriterFactory valueWriterFactory;
+    private int valueSequenceCounter;
+    private List<FlatMapKeyNodeWriter> keyNodes;
+    private boolean closed;
+    private Long2ObjectMap<FlatMapKeyNodeWriter> keyToKeyNode = new Long2ObjectOpenHashMap<>();
+
+    public FlatMapNumericKeyBlockWriter(
+            int valueColumnId,
+            ColumnWriterOptions columnWriterOptions,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            MetadataWriter metadataWriter,
+            Type type,
+            OrcType orcKeyType,
+            ColumnWriterFactory valueWriterFactory,
+            List<FlatMapKeyNodeWriter> keyNodes)
+    {
+        this.valueColumnId = valueColumnId;
+        MapType mapType = (MapType) type;
+        this.keyType = mapType.getTypeParameters().get(0);
+        this.valueType = mapType.getTypeParameters().get(1);
+        this.orcKeyType = orcKeyType;
+
+        this.columnWriterOptions = columnWriterOptions;
+        this.dwrfEncryptor = dwrfEncryptor;
+        this.metadataWriter = metadataWriter;
+        this.valueWriterFactory = valueWriterFactory;
+        this.keyNodes = keyNodes;
+
+        this.valueSequenceCounter = 1;
+    }
+
+    public Long getLongValue(Block keysBlock, int position)
+    {
+        Long keyAsLong;
```

Comment:
This implementation is very inefficient, for each value it is creating lot of garbage.

Use of SqlFunctionProperties is also not required.